velox/substrait/SubstraitToVeloxPlan.cpp

```diff
@@ -1243,70 +1244,91 @@ void SubstraitVeloxPlanConverter::constructSubfieldFilters(
   using RangeType = typename RangeTraits<KIND>::RangeType;
   using MultiRangeType = typename RangeTraits<KIND>::MultiRangeType;
 
+  if (!filterInfo->isInitialized()) {
+    return;
+  }
+
+  uint32_t rangeSize = std::max(
+      filterInfo->lowerBounds_.size(), filterInfo->upperBounds_.size());
+  bool nullAllowed = filterInfo->nullAllowed_;
+
   // Handle 'in' filter.
   if (filterInfo->valuesVector_.size() > 0) {
     setInFilter<KIND>(
-        filterInfo->valuesVector_,
-        filterInfo->nullAllowed_,
-        inputName,
-        filters);
+        filterInfo->valuesVector_, nullAllowed, inputName, filters);
     // Currently, In cannot coexist with other filter conditions
-    // due to multirange is in 'OR' relation but 'AND' can be needed.
+    // due to multirange is in 'OR' relation but 'AND' is needed.
+    VELOX_CHECK(
+        rangeSize == 0,
+        "LowerBounds or upperBounds conditons cannot be supported after IN filter.");
     VELOX_CHECK(
-        filterInfo->lowerBounds_.size() == 0,
-        "Other conditons cannot be supported.");
+        nullAllowed, "Null filtering cannot be supported after IN filter.");
     VELOX_CHECK(
-        filterInfo->upperBounds_.size() == 0,
-        "Other conditons cannot be supported.");
+        !filterInfo->notValue_.has_value(),
+        "Not equal cannot be supported after IN filter.");
     return;
   }
 
   // Construct the Filters.
   std::vector<std::unique_ptr<FilterType>> colFilters;
+
+  // Handle not(equal) filter.
+  if (filterInfo->notValue_) {
+    variant notVariant = filterInfo->notValue_.value();
+    createNotEqualFilter<KIND, FilterType>(
+        notVariant, filterInfo->nullAllowed_, colFilters);
+    // Currently, Not-equal cannot coexist with other filter conditions
+    // due to multirange is in 'OR' relation but 'AND' is needed.
+    VELOX_CHECK(
+        rangeSize == 0,
+        "LowerBounds or upperBounds conditons cannot be supported after not-equal filter.");
+    filters[common::Subfield(inputName)] =
+        std::make_unique<MultiRangeType>(std::move(colFilters), nullAllowed);
+    return;
+  }
+
+  // Handle null filtering.
+  if (rangeSize == 0 && !nullAllowed) {
+    std::unique_ptr<common::IsNotNull> filter =
+        std::make_unique<common::IsNotNull>();
+    filters[common::Subfield(inputName)] = std::move(filter);
```

Comment:
The IsNotNull pushdown brings ~17s performance regression for TPC-H, but it is functionally needed.